plugins/thunderbolt-power/fu-plugin-thunderbolt-power.c

```diff
@@ -0,0 +1,273 @@
+
+#include <errno.h>
+#include <fcntl.h>
+#include <gio/gio.h>
+#include <glib.h>
+#include <glib/gstdio.h>
+#include <gudev/gudev.h>
+
+#include "fu-plugin-vfuncs.h"
+#include "fu-device-metadata.h"
+
+#ifndef HAVE_GUDEV_232
+G_DEFINE_AUTOPTR_CLEANUP_FUNC(GUdevDevice, g_object_unref)
+#endif
+
+/* empirically mesaured amount of time for the TBT device to come and go */
+#define TBT_NEW_DEVICE_TIMEOUT	2
+
+struct FuPluginData {
+	GUdevClient   *udev;
+	gchar         *force_path;
+	gboolean       needs_forcepower;
+};
+
+static void
+fu_plugin_thunderbolt_power_get_path (FuPlugin *plugin)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	g_autoptr(GList) devices = NULL;
+	const gchar *basepath;
+	const gchar *driver;
+
+	devices = g_udev_client_query_by_subsystem (data->udev, "wmi");
+	for (GList* l = devices; l != NULL; l = l->next) {
+		g_autofree gchar *built_path = NULL;
+		g_autoptr(GUdevDevice) device = l->data;
+
+		/* only supports intel-wmi-thunderbolt for now */
+		driver = g_udev_device_get_driver (device);
+		if (g_strcmp0 (driver, "intel-wmi-thunderbolt") != 0)
+			continue;
+
+		/* check for the attribute to be loaded */
+		basepath = g_udev_device_get_sysfs_path (device);
+		if (basepath == NULL)
+			continue;
+		built_path = g_build_path ("/", basepath,
+					   "force_power", NULL);
+		if (g_file_test (built_path, G_FILE_TEST_IS_REGULAR)) {
+			data->force_path = g_steal_pointer (&built_path);
+			g_debug ("Detected force power support at %s",
+				 data->force_path);
+			return;
+		}
+	}
+
+	/* driver went away */
+	if (data->force_path != NULL)
+		g_free (data->force_path);
+}
+
+static gboolean
+fu_plugin_thunderbolt_power_supported (FuPlugin *plugin)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	return (data->force_path != NULL);
+}
+
+static gboolean
+fu_plugin_thunderbolt_power_set (FuPlugin *plugin, gboolean enable,
+				 GError **error)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	gint fd;
+	gint ret;
+
+	if (!fu_plugin_thunderbolt_power_supported (plugin)) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "unable to set power to %d (missing kernel support)",
+			     enable);
+		return FALSE;
+	}
+	g_debug ("Setting force power to %d", enable);
+	fd = g_open (data->force_path, O_WRONLY);
+	if (fd == -1) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INTERNAL,
+			     "failed to open %s", data->force_path);
+		return FALSE;
+	}
+	ret = write (fd, enable ? "1" : "0", 1);
+	if (ret < 1) {
+		g_set_error (error, G_IO_ERROR,
+			     g_io_error_from_errno (errno),
+			     "could not write to force_power': %s",
+			     g_strerror (errno));
+		g_close (fd, NULL);
+		return FALSE;
+	}
+	return g_close (fd, error);
+}
+
+static gboolean
+fu_plugin_thunderbolt_power_reset_cb (gpointer data)
+{
+	if (!fu_plugin_thunderbolt_power_set (FU_PLUGIN (data), FALSE, NULL))
+		g_warning ("failed to reset thunderbolt power");
+	return FALSE;
```

Comment:
Always `FALSE`? I'd think always `TRUE` if it must return anything or using the result of `set`.

Reply:
For the `g_timeout` to go away on the reset callback you have to return `FALSE`.  It's for `g_timeout`'s to call repeatedly on an interval instead of just once when you return `TRUE`.

Reply:
Oh, sorry. Next time I'll look it up myself before asking...
Thanks!

Reply:
Note, you can use `G_SOURCE_REMOVE` which makes it a bit clearer in this kind of situation. I'm bad at using it myself.